Parser/pegen/pegen.c

```diff
@@ -756,24 +756,15 @@ _PyPegen_expect_token(Parser *p, int type)
 expr_ty
 _PyPegen_expect_soft_keyword(Parser *p, const char *keyword)
 {
-    if (p->mark == p->fill) {
-        if (_PyPegen_fill_token(p) < 0) {
-            p->error_indicator = 1;
-            return NULL;
-        }
-    }
-    Token *t = p->tokens[p->mark];
-    if (t->type != NAME) {
-        return NULL;
-    }
-    char* s = PyBytes_AsString(t->bytes);
-    if (!s) {
+    expr_ty res = _PyPegen_name_token(p);
+    if (!res) {
         return NULL;
     }
-    if (strcmp(s, keyword) != 0) {
+    const char *s = PyUnicode_AsUTF8(res->v.Name.id);
+    if (!s || strcmp(s, keyword) != 0) {
```

Comment:
Whoa. If the strcmp fails, the error indicator shouldn't be set. That's just failure to find the keyword. *And*, it should rest the mark, since we're not recognizing a token in that case.

Reply:
> If the strcmp fails, the error indicator shouldn't be set.

That's right!

> And, it should rest the mark, since we're not recognizing a token in that case.

I thought about it and I went with not resetting the `mark`, since if this rule returns NULL, then the `if` in the generated alternative immediately fails and the code there resets the mark to be `p->mark`. I don't really know, if that's okay, since it makes it a little bit harder to understand the code, but it should work. What do you think?